HACKING.md

```diff
@@ -37,15 +37,13 @@ Note that `tools/cockpit.spec` is a template filled in by
 `tools/gen-spec-dependencies`, and cannot be directly parsed by dnf.
 The following should work in a fresh Git clone:
 
-    $ sudo dnf install dnf-utils
-    $ TEMPFILE=$(mktemp -u --suffix=.spec) && \
-      sed 's/%{npm-version:.*}/0/' tools/cockpit.spec >$TEMPFILE && \
-      sudo dnf builddep --spec $TEMPFILE && \
-      rm $TEMPFILE
+    $ sudo dnf install dnf-utils python-srpm-macros
```

Comment:
The previous approach would not actually change any files in checkout. When I do your approach then `git diff` contains bunch of changes.
Since one would do this once, it is fine I guess. By any chance can you explain why the previous approach did not work for you and you had to use `tools/gen-spec-dependencies`?

Reply:
I thought I should run those commands, but now I see the JS is not needed for building. Right, I think this confused me:
> To actually build the Cockpit binaries themselves from source (including to run the integration tests locally), you will need additional header files and other components. **Check tools/cockpit.spec for the concrete Fedora build dependencies.**

> Note that tools/cockpit.spec is a template filled in by tools/gen-spec-dependencies, and cannot be directly parsed by dnf. The following should work in a fresh Git clone:

May I suggest a wording change:

To actually build the Cockpit binaries themselves from source, including to run the integration tests locally, you will need additional header files and other components. tools/cockpit.spec contains the concrete Fedora build dependencies. However, it is a template, and so cannot be directly parsed by dnf. It is modified by tools/gen-spec-dependencies when making the dist target (is that true?). As the JS/node/npm/etc/? dependencies are only used for packaging (or whatever the true reason is), to install Fedora build dependencies, the following should work in a fresh Git clone:

The original sounds like it wants me actually check/look at the contents of the file. Then straight away mentions using gen-spec-dependencies. The above might be more clear. I think also the code formatting could be dropped around tools/gen-spec-dependencies and tools/cockpit.spec, as it put it's on par with the other commands in this document that one should run.

Reply:
Never mind about the code formatting (backticks). I should wait till read more docs before making such a suggestion. I did look over these `https://github.com/cockpit-project/cockpit/wiki/Cockpit-Coding-Guidelines`, but maybe it's just an unspoken convention.

Reply:
> It is modified by tools/gen-spec-dependencies when making the dist target (is that true?)

It is. Or you can just say "when doing official builds".

> As the JS/node/npm/etc/? dependencies are only used for packaging (or whatever the true reason is)

There is no "JS" per se. Node and npm are used only to put all the things together and the result that is produced does not need these anymore.

> I think also the code formatting could be dropped around tools/gen-spec-dependencies and tools/cockpit.spec, as it put it's on par with the other commands in this document that one should run.

Lets not do that. It is a good practice to format files and commands this way. It makes it much easier to read as you know what you need to parse.

Reply:
I hope that is clearer now. There were also some more packages that needed to be installed for the tests. It took a while to understand what my problem was. I think it may have been IOMMU related, but not sure, as I have no problems with other nested OSes on the same hardware. So I'm not sure if it IOMMU is the problem. I'd like to improve the docs in this regard also.